src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -141,23 +141,37 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
 
             async ValueTask<(Solution? solution, T result)> TryFastGetSolutionAndRunAsync()
             {
-                Solution solution;
+                Solution solutionTemp;
                 using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
                 {
                     if (_lastRequestedPrimaryBranchSolution.checksum == solutionChecksum)
                     {
-                        solution = _lastRequestedPrimaryBranchSolution.solution;
+                        solutionTemp = _lastRequestedPrimaryBranchSolution.solution;
                     }
                     else if (_lastRequestedAnyBranchSolution.checksum == solutionChecksum)
                     {
-                        solution = _lastRequestedAnyBranchSolution.solution;
+                        solutionTemp = _lastRequestedAnyBranchSolution.solution;
                     }
                     else
                     {
                         return default;
                     }
                 }
 
+                // We successfully found the solution.  However, we still want to ensure it's in the
+                // _solutionChecksumToLazySolution map.  This ensures that as long as the call to `implementation` is in
+                // progress subsequent calls with this same checksum will be able to reuse this solution. If we did not
+                // do this, an intervening edit might cause _lastRequestedXXX to change, and subsequence calls to this
+                // same solutionChecksum would then not find the item either there, or in the
+                // _solutionChecksumToLazySolution map.
+
+                var refCountedLazySolution = await GetLazySolutionAsync(
+                    solutionChecksum,
+                    _ => Task.FromResult(solutionTemp),
+                    cancellationToken).ConfigureAwait(false);
+                await using var configuredAsyncDisposable = refCountedLazySolution.ConfigureAwait(false);
+                var solution = await refCountedLazySolution.Target.Task.WithCancellation(cancellationToken).ConfigureAwait(false);
```

Comment:
this is the meat of hte change.  with GetLazySolutionAsync becoming a helper function so it can be used here.

Reply:
discussed offline if maybe having the last fast solution be a part of the _solutionChecksumToLazySolution (and ref counted) might be slightly cleaner.